### PR TITLE
pkg: add local packages hash to lockdir

### PR DESCRIPTION
**Pull request by gridbugs**
If the package definitions in dune-project are changed after a lockdir is generated then the lockdir may no longer contain a valid solution for the local packages in the project. To help detect when this has happened this change adds a new field to the lockdir containing a hash of all the local package metadata used for generating it. When the local package metadata would change in a way that may affect the solution we can detect this by computing a new hash of the local packages and comparing it against the one stored in the lockdir.

### CODE REVIEW

**Comment by gridbugs** [comment]
Making this PR as a strawman to start a conversation about whether this is the right way to detect when dune-project becomes out of sync with the lockdir. Now that I've thought a little more about this in the process of making this change my preference for solving this problem would be to check that all dependencies of local packages can be satisfied from the lockdir. That would have the benefit of not needing to store any more information in the lockdir, would allow us to produce richer error messages than just "the hashes don't match", and also detect the case where a user has tampered with their lockdir (which hashing local packages obviously won't catch).

**Comment by rgrinberg** [comment]
> . Now that I've thought a little more about this in the process of making this change my preference for solving this problem would be to check that all dependencies of local packages can be satisfied from the lockdir. That would have the benefit of not needing to store any more information in the lockdir, would allow us to produce richer error messages than just "the hashes don't match", and also detect the case where a user has tampered with their lockdir (which hashing local packages obviously won't catch).

This doesn't really help when the user updates their constraints, right?

**Comment by Leonidas-from-XIV** [comment]
Personally I think creating some kind of checksum of the input to the solver and printing a warning that the lock folder might be out of date seems fine - a simple and understandable solution that is easy to explain to users.

> Now that I've thought a little more about this in the process of making this change my preference for solving this problem would be to check that all dependencies of local packages can be satisfied from the lockdir.

How would you do this? The obvious solution to me to run the solver with only the currently selected packages as candidates works but I worry that running the solver all the time would be a bit time-consuming and not necessarily correct as the packages metadata in the repo might have changed (e.g. a package was marked as non-available etc).

**Comment by Leonidas-from-XIV** [comment]
Generally: a good idea. I am always annoyed when I get my `dune-project` and the generated opam files out of sync by accident and with lock files this is even easier as you need to lock specifically.

Bike shedding: I find the naming "local package hash" confusing as as an user I would call this "dependency hash", as this describes the dependencies that I have declared.

Also, if we use such a hash (or any method of checking that the lock dirs and `dune-project` are in sync), I don't think it should be a separate command, as that basically requires users to check manually - and if they remember to do, they could also remember to re-lock. It should be happening any time dune would use the lock dir as this is the place where things go wrong. This of course means that any such check would need to be fast enough to be run automatically.

**Comment by gridbugs** [comment]
I've updated this PR so that it only hashes the set of non-local dependencies of local packages. This way certain operations such as adding a package which only depends on existing dependencies (including their constraints) doesn't affect the hash and so won't require regenerating the lockdir.

**Comment by gridbugs** [comment]
> I don't think it should be a separate command, as that basically requires users to check manually

Agreed. I've done it this way for now as we currently don't have any operations that require the lockdir and local packages to be in sync. The first such operation would be https://github.com/ocaml/dune/pull/9095 which is what motivated me to make this PR.

**Comment by gridbugs** [comment]
> > Now that I've thought a little more about this in the process of making this change my preference for solving this problem would be to check that all dependencies of local packages can be satisfied from the lockdir.
> 
> How would you do this? The obvious solution to me to run the solver with only the currently selected packages as candidates works but I worry that running the solver all the time would be a bit time-consuming and not necessarily correct as the packages metadata in the repo might have changed (e.g. a package was marked as non-available etc).

There's enough information already in the lockdir to compute the transitive dependency closure of all local packages and detect when a dependency is absent. There's no need to run the solver again.

**Comment by gridbugs** [comment]
Converted to draft while https://github.com/ocaml/dune/pull/9172 is in review which validates lockdirs against local packages by analyzing dependencies of local packages and checking if they are contained in the lockdir. That change is needed regardless of the dependency hash in order to gracefully handle the case where files are added/removed/changed in the lockdir in a way which invalidates it w.r.t the local packages (which should be rare but we still need to handle it). Said PR steps on this one's toes a bit so I'll wait for the dust to settle before dealing with all the merge conflicts.

**Comment by gridbugs** [comment]
Updated on top of my recent changes to lockdir validation. This is now ready for review.

File: src/dune_lang/package_constraint.ml
Comment (Leonidas-from-XIV) [inline src/dune_lang/package_constraint.ml:29]: What is the semantics of this pattern match? Maybe writing down a comment on the ordering would be helpful, since I don't think there's a natural ordering for comparison operators.

File: src/dune_pkg/dune
Comment (Leonidas-from-XIV) [inline src/dune_pkg/dune:25]: Careful, the `sha` package is only vendored because of OPAM, I would be a bit hesitant to just use it directly in other code and probably go for `Digest` instead.

Or at least decide what kinds of hashes we want to use in Dune and deliberately vendor that hash library.

File: src/dune_pkg/local_package.ml
Comment (Leonidas-from-XIV) [inline src/dune_pkg/local_package.ml:12]: Another thing to be aware of is how to deal with hash migration. Maybe a scheme where the hash is prefixed by its algorithm like OPAMs `sha1=deadcafeb0ba` would make sense in places where the user is exposed to the hash?

File: src/dune_pkg/local_package.mli
Comment (Leonidas-from-XIV) [inline src/dune_pkg/local_package.mli:41]: I think it is a good idea to distinguish this, but wouldn't `Dependency_hash.t option` suffice? The `None` case is mostly used if something can't be sensibly constructed, exactly like `Empty` here, but with the added benefit of `Option.O`.

File: src/dune_pkg/dune
Comment (gridbugs) [reply src/dune_pkg/dune]: switched to use `Digest`

File: src/dune_pkg/local_package.ml
Comment (gridbugs) [reply src/dune_pkg/local_package.ml]: changed the format to `md5=<hash>`

Comment (rgrinberg) [reply src/dune_pkg/local_package.ml]: I'm sorry I'm a bit late to the party, but I think the hash algorithm should be controlled by the version of the lock directory set in `lock.dune`. The whole hash thing is a complete implementation detail of dune and there's no need to expose any details to the user.

**Review by rgrinberg — APPROVED** [review]

**Pull request merged**
